src/Interpreters/SyntaxAnalyzer.cpp

```diff
@@ -373,6 +374,126 @@ void optimizeOrderBy(const ASTSelectQuery * select_query)
+
+    if (function && function->isStateful())
+        return true;
+
+    if (ast_function->arguments)
+    {
+        ASTs args = ast_function->arguments->children;
+        for (const auto & elem : args)
+        {
+            if (const auto arg = elem->as<ASTFunction>())
+            {
+                if (arg && isASTFunctionStateful(arg, context))
+                    return true;
+            }
+        }
+    }
+
+    return false;
+}
+
+/// Removes duplicate ORDER BY from subqueries.
+void optimizeDuplicateOrderByFromSubqueries(const ASTPtr & current_ast, const Context & context)
+{
+    if (!current_ast)
+        return;
+
+    if (current_ast->getID() == "SelectQuery")
+    {
+        auto select_query = current_ast->as<ASTSelectQuery>();
+
+        if (select_query->orderBy() && !select_query->limitBy() && !select_query->limitByOffset() &&
+            !select_query->limitByLength() && !select_query->limitLength() && !select_query->limitOffset())
+        {
+            select_query->setExpression(ASTSelectQuery::Expression::ORDER_BY, nullptr);
+        }
+    }
+    else
+    {
+        for (const auto & elem : current_ast->children)
+        {
+            optimizeDuplicateOrderByFromSubqueries(elem, context);
+        }
+    }
+}
+
+/// Checkss if duplicate ORDER BY from subqueries can be erased.
+void optimizeDuplicateOrderBy(const ASTPtr & current_ast, const Context & context)
+{
+    for (const auto & elem : current_ast->children)
+        optimizeDuplicateOrderBy(elem, context);
+
+    auto select_query = current_ast->as<ASTSelectQuery>();
+
+    if (!select_query)
+        return;
+
+    if (select_query->orderBy() || select_query->groupBy())
+    {
+        const auto & expression_list = select_query->select();
+
+        for (const auto & ast_function : expression_list->children)
+        {
+            std::cerr << ast_function;
+            auto function = ast_function->as<ASTFunction>();
+            if (function && isASTFunctionStateful(function, context))
+                return;
+        }
+
+        optimizeDuplicateOrderByFromSubqueries(select_query->tables(), context);
+    }
+}
+
+/// Removes duplicate DISTINCT from subquries.
+void optimizeDuplicateDistinct(const ASTPtr & current_ast,
+                               bool & is_distinct,
+                               std::vector<String> & last_ids)
+{
+    if (!current_ast)
+        return;
+
+    for (auto & child : current_ast->children)
+    {
+        optimizeDuplicateDistinct(child, is_distinct, last_ids);
+    }
+
+    if (current_ast->getID() == "SelectQuery")
+    {
+        auto select_query = current_ast->as<ASTSelectQuery>();
+
+        if (select_query->distinct)
+        {
+            auto & expression_list = select_query->select();
+            std::vector<String> current_ids;
+
+            for (const auto & id : expression_list->children)
+                current_ids.push_back(id->getColumnName());
```

Comment:
The asterisks does not work, i.e. `select distinct * from (select distinct * from numbers(10))`